qiskit/quantum_info/states/utils.py

```diff
@@ -155,3 +154,52 @@ def _funm_svd(matrix, func):
     unitary1, singular_values, unitary2 = la.svd(matrix)
     diag_func_singular = np.diag(func(singular_values))
     return unitary1.dot(diag_func_singular).dot(unitary2)
+
+
+def partial_transpose(state, qargs):
+    """Return partially transposed density matrix.
+
+    Args:
+        state (DensityMatrix): the input state.
+        qargs (list): The subsystems to be transposed.
+
+    Returns:
+        DensityMatrix: The partially transposed density matrix.
+
+    Raises:
+        QiskitError: if input state is invalid.
+        QiskitError: if indices of subsystems are invalid
+
+    """
+    state = _format_state(state, validate=False)
+    n = len(state.dims())
+    lst = np.zeros(2**n, int)
+    for i in range(2**n):
+        x = 0
+        for k in qargs:
+            x = x + (((i >> (k)) % 2) * 2**k)
+        lst[i] = x
+    if not set(qargs).issubset(set(np.arange(n))):
+        raise QiskitError("Indices of subsystems to be transposed are invalid")
+    ptden = np.empty((2**n, 2**n), complex)
+    ptden[:] = np.nan
+    if isinstance(state, Statevector):
+        state = np.array(state)
+        for i in range(2**n):
+            for j in range(2**n):
+                if math.isnan(ptden[i, j]):
+                    x = i - lst[i] + lst[j]
+                    y = j - lst[j] + lst[i]
+                    ptden[i, j] = state[x] * np.conjugate(state[y])
+                    ptden[x, y] = state[i] * np.conjugate(state[j])
+    else:
+        state = np.array(state)
+        for i in range(2**n):
+            for j in range(2**n):
+                if math.isnan(ptden[i, j]):
```

Comment:
Could this loop be made more efficient by only visiting entries that we know need to be populated, or maybe transposing sub-blocks of the matrix all at once? As it is now, this scales as ``2 ^ (2n)`` which is going to be very expensive.

Reply:
We will try to optimize it.

Reply:
For most operators, I would imagine you can do this in vectorised Numpy by casting `state.data` (a Numpy array) to a multi-dimensional tensor (e.g. using `np.reshape(state.data, state._op_shape.tensor_shape)` or something like that), calling `np.transpose` as appropriate, then restoring it to be a 2D array.  If that's correct, then that should be much faster than looping in Python space.

`quantum_info.Operator` also isn't guaranteed to be square.  I don't think the partial trace is reasonably defined for tensor-product spaces where the number of input and output spaces are unequal, but it could in theory work for things like a mapping of a `(2, 3)` Hilbert space to a `(1, 4)` one.  Probably best not to worry about that, though, and just error out if the `dims_r` of the `OpShape` aren't equal to the `dims_l`.

Reply:
@jakelishman This helps. Thank you.